webhooks/pod_webhook.go

```diff
@@ -230,57 +221,116 @@ func (m *PodMutator) injectSidecar(
 	return json.Marshal(pod)
 }
 
-func (m *PodMutator) handleHttpProvider(sidecar *corev1.Container, source v1alpha1.Source) {
-	// append args
-	sidecar.Args = append(
-		sidecar.Args,
-		"--uri",
-		source.Source,
-	)
-	if source.HttpSyncBearerToken != "" {
-		sidecar.Args = append(
-			sidecar.Args,
-			"--bearer-token",
-			source.HttpSyncBearerToken,
-		)
+// buildSources builds types.SourceConfig collection to be used by the sidecar.
+func (m *PodMutator) buildSources(ctx context.Context, flagSourceConfig *v1alpha1.FlagSourceConfigurationSpec,
+	pod *corev1.Pod, sidecar *corev1.Container) ([]types.SourceConfig, error) {
+
+	var sourceCfgCollection []types.SourceConfig
+
+	for _, source := range flagSourceConfig.Sources {
+		if source.Provider == "" {
+			source.Provider = flagSourceConfig.DefaultSyncProvider
+		}
+
+		var sourceCfg types.SourceConfig
+		var err error
+
+		switch {
+		case source.Provider.IsKubernetes():
+			sourceCfg, err = m.toKubernetesConfig(ctx, pod, source)
+			if err != nil {
+				return []types.SourceConfig{}, err
+			}
+		case source.Provider.IsFilepath():
+			sourceCfg, err = m.handleFilepathProvider(ctx, pod, sidecar, source)
+			if err != nil {
+				return []types.SourceConfig{}, err
+			}
+		case source.Provider.IsHttp():
+			sourceCfg = m.toHttpProviderConfig(source)
+		case source.Provider.IsGrpc():
+			sourceCfg = m.toGrpcProviderConfig(source)
+		default:
+			return []types.SourceConfig{}, fmt.Errorf("unrecognized sync provider in config: %s", source.Provider)
+		}
+
+		sourceCfgCollection = append(sourceCfgCollection, sourceCfg)
+
+	}
+
+	return sourceCfgCollection, nil
+}
+
+// toHttpProviderConfig generate types.SourceConfig for http provider
+func (m *PodMutator) toHttpProviderConfig(source v1alpha1.Source) types.SourceConfig {
+	return types.SourceConfig{
+		URI:         source.Source,
+		Provider:    string(v1alpha1.SyncProviderHttp),
+		BearerToken: source.HttpSyncBearerToken,
 	}
 }
 
-func (m *PodMutator) handleKubernetesProvider(ctx context.Context, pod *corev1.Pod, sidecar *corev1.Container, source v1alpha1.Source) error {
+// toGrpcProviderConfig generate types.SourceConfig for grpc provider
+func (m *PodMutator) toGrpcProviderConfig(source v1alpha1.Source) types.SourceConfig {
+	return types.SourceConfig{
+		URI:        source.Source,
```

Comment:
the prefix should be added here, either grpc or grpcs dependent on the certPath field

Reply:
Now that the provider type is being explicitly given can the notion of grpc/grpcs be retired entirely and inferred from the provision of certificates?

Reply:
There's one corner case that we cannot handle alone with the presence of a certificate - grpcs can work without a certificate by utilizing system certificates.

Besides, just found out the mixed usage of `file` vs `filepath` terminology usage in flagd vs OFO [1]. This shows why we should not attempt to do a lot of modifications from the OFO side as OFO & flagd can diverge as they are independent repositories 

[1] - https://github.com/open-feature/open-feature-operator/blob/ce092b6c6160da0ad51b1b9aaf443152e6f6b833/webhooks/pod_webhook.go#L373

Reply:
TL;DR: Lets remove the requirement for the grpc/grpcs prefixes in flagd by introducing a boolean `useTLS` configuration. Otherwise we will always need to prefix host:port with `grpc(s)://` which is something we made up for flagd's `--uri` flag


With the current configuration. if I was to pass:
```
sources:
 - provider: grpc
   uri: 127.0.0.1:8080
```
flagd would error on startup because my `host:port` does not start with 'grpc://', this is not an issue related to this PR, but I do think we should resolve it before we make these changes:

if we extend the config object to include `useTLS: bool` we no longer require the prefix which is currently being [enforced here](https://github.com/open-feature/flagd/blob/7da78e9c48ccdfd9a8f682a78a0c8d63216da79d/core/pkg/sync/grpc/grpc_sync.go#L238). 

We would still of course use this prefix for a `--uri` flag, but it feels like inappropriate duplication in the configuration object.

Now we can pass flagd grpc configurations as follows:
```
sources:
 - provider: grpc
   uri: 127.0.0.1:8080
   useTLS: true
```
Adding a certificate path where required, but also allowing us to use system certificates if we require it.

We can then add this field to the `sources` definition of the `FlagSourceConfiguration` CRD to keep everything clean on the OFO side + include these changes in this PR


as for Kube / file sync we [regex replace these prefixes](https://github.com/open-feature/flagd/blob/7314feea8c7bc90aac0528a9e1be0759a7a60c15/core/pkg/runtime/from_config.go#L189) if they are present before starting the syncs so they can also be removed. I think this is a much cleaner approach as we no longer are reliant on the prefixes we introduced for the `--uri` flags

Reply:
We could export these from flagd, as constants.

Reply:
> TL;DR: Lets remove the requirement for the grpc/grpcs prefixes in flagd by introducing a boolean `useTLS` configuration. Otherwise we will always need to prefix host:port with `grpc(s)://` which is something we made up for flagd's `--uri` flag
> 
> With the current configuration. if I was to pass:
> 
> ```
> sources:
>  - provider: grpc
>    uri: 127.0.0.1:8080
> ```
> 
> flagd would error on startup because my `host:port` does not start with 'grpc://', this is not an issue related to this PR, but I do think we should resolve it before we make these changes:
> 
> if we extend the config object to include `useTLS: bool` we no longer require the prefix which is currently being [enforced here](https://github.com/open-feature/flagd/blob/7da78e9c48ccdfd9a8f682a78a0c8d63216da79d/core/pkg/sync/grpc/grpc_sync.go#L238).
> 
> We would still of course use this prefix for a `--uri` flag, but it feels like inappropriate duplication in the configuration object.
> 
> Now we can pass flagd grpc configurations as follows:
> 
> ```
> sources:
>  - provider: grpc
>    uri: 127.0.0.1:8080
>    useTLS: true
> ```
> 
> Adding a certificate path where required, but also allowing us to use system certificates if we require it.
> 
> We can then add this field to the `sources` definition of the `FlagSourceConfiguration` CRD to keep everything clean on the OFO side + include these changes in this PR
> 
> as for Kube / file sync we [regex replace these prefixes](https://github.com/open-feature/flagd/blob/7314feea8c7bc90aac0528a9e1be0759a7a60c15/core/pkg/runtime/from_config.go#L189) if they are present before starting the syncs so they can also be removed. I think this is a much cleaner approach as we no longer are reliant on the prefixes we introduced for the `--uri` flags

I strongly agree with all the above.

Reply:
> > Besides, just found out the mixed usage of file vs filepath terminology usage in flagd vs OFO [1]. This shows why we should not attempt to do a lot of modifications from the OFO side as OFO & flagd can diverge as they are independent repositories
> 
> We could export these from flagd, as constants.

flagd have a dependency on OFO. So we might need to avoid circular dependency situation by adding flagd dependency to OFO. 

At least this was the aim with this `SourceConfig` [1]

[1] - https://github.com/open-feature/open-feature-operator/blob/1993f6c14eabefa22c1ef82d45c97d8a14d90d5c/pkg/types/SourceConfig.go

Reply:
I am not against adding one more field `useTLS` and using that for grpc. But more configurations can lead to confusion. 

And do you think we need to change flagd? Because this can be handled at OFO side similar to how we handle file & k8s. 

Otherwise, we need to refactor flagd and document the new behavior. Which also includes the introduction of a new source argument for`useTLS`

Reply:
> And do you think we need to change flagd? Because this can be handled at OFO side similar to how we handle file & k8s.

IMO we should introduce these changes into flagd first

Reply:
We could have already seen this when I worked on this PR - https://github.com/open-feature/flagd/pull/532

Reply:
Also, @james-milligan  was there any intention to allow both prefixed and non-prefixed resources URIs at the first place ? I am refering to here [1] 

[1] - https://github.com/open-feature/flagd/blob/dc9cb5fa2005630fd68c7bc0a63a2b056430ff0d/core/pkg/runtime/from_config.go#L189-L192

Reply:
Yes, personally I think delaying this a bit for the required changes in flagd is fine. That's my opinion at least.

I also _think_ (hope) that there shouldn't be too much flux here that you will get a ton of conflicts if you put this on ice while making the flagd change.

> We could have already seen this when I worked on this PR - https://github.com/open-feature/flagd/pull/532

Ya, I think we missed this one! It's sometimes hard to see things like this until they all start to come together, but I think what we are proposing is a good improvement. I want to rely on the invalid `grpc(s)` scheme as little as possible. It's always worried me, TBH.

Reply:
Thanks for the discussion :) 

I will close this PR as it will go stale (ex:- conflicts) for now and rework once we have flagd changes ready.